src/main/kotlin/io/libp2p/security/noise/NoiseXXSecureChannel.kt

```diff
@@ -25,9 +27,11 @@ import org.apache.logging.log4j.core.config.Configurator
 import spipe.pb.Spipe
 import java.util.concurrent.CompletableFuture
 
-class NoiseXXSecureChannel(val localKey: PrivKey, val localDHState: DHState, val remoteDHState: DHState, val role: Int) :
+open class NoiseXXSecureChannel(val localKey: PrivKey, val privateKey25519: ByteArray) :
```

Comment:
Does Noise really require two private keys?

Reply:
One private key is used to verify the peer identity, the other private key is the Noise 25519 key.

Reply:
Trying to dig into this thread now https://github.com/libp2p/specs/issues/195
Yep, the question doesn't look straightforward 

I believe your Telegram channel comment was closely related

> A few points from my end from my very quick perusal of that thread. Take what I say with a grain of salt.
> - As far as I can tell, but maybe I just couldn't find it or didn't recognize some API call, rust-libp2p's current implementation doesn't incorporate any ephemeral keys. This doesn't match Noise protocols like IN, IK, IX (pointed out as not supported on the thread), nor XX (mentioned to be supported in that thread). i.e., I didn't see any DH actions based on ephemeral keys. 
> - As well, as far as I know, transmitting packets requires the use of chaining keys as communication continues, which I couldn't find (maybe I missed it) in the rust-libp2p implementation. 
> - reusing Curve25519 keys in ed25519 or vice-versa is not an easy option. The byte value can be used as an origin private key, but the public keys will be different and will be used for different purposes (ed25519 is for signing and Curve25519 is not?). A (minor) side-effect is that the key interface in jvm-libp2p (which has interface methods for signing/verifying using private/public keys) are not to be implemented.